WebDriverAgentLib/Utilities/FBImageIOScaler.m

```diff
@@ -0,0 +1,127 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import "FBImageIOScaler.h"
+#import <ImageIO/ImageIO.h>
+#import <MobileCoreServices/MobileCoreServices.h>
+#import "FBLogger.h"
+
+static const CGFloat FBMinScalingFactor = 0.01f;
+static const CGFloat FBMaxScalingFactor = 1.0f;
+static const CGFloat FBMinCompressionQuality = 0.0f;
+static const CGFloat FBMaxCompressionQuality = 1.0f;
+
+@interface FBImageIOScaler ()
+
+@property (nonatomic) NSData *nextImage;
+@property (nonatomic, readonly) NSLock *nextImageLock;
+@property (nonatomic, readonly) dispatch_queue_t scalingQueue;
+
+@end
+
+@implementation FBImageIOScaler
+
+- (id)init
+{
+  self = [super init];
+  if (self) {
+    _nextImageLock = [[NSLock alloc] init];
+    _scalingQueue = dispatch_queue_create("image.scaling.queue", NULL);
+  }
+  return self;
+}
+
+- (void)submitImage:(NSData *)image scalingFactor:(CGFloat)scalingFactor compressionQuality:(CGFloat)compressionQuality completionHandler:(void (^)(NSData *))completionHandler {
+  [self.nextImageLock lock];
+  if (self.nextImage != nil) {
+    [FBLogger verboseLog:@"Discarding screenshot"];
+  }
+  scalingFactor = MAX(FBMinScalingFactor, MIN(FBMaxScalingFactor, scalingFactor));
+  compressionQuality = MAX(FBMinCompressionQuality, MIN(FBMaxCompressionQuality, compressionQuality));
+  self.nextImage = image;
+  [self.nextImageLock unlock];
+
+  dispatch_async(self.scalingQueue, ^{
+    [self.nextImageLock lock];
+    NSData *next = self.nextImage;
+    self.nextImage = nil;
+    [self.nextImageLock unlock];
+    if (next == nil) {
+      return;
+    }
+    NSData *scaled = [self scaledImageWithImage:next
+                                  scalingFactor:scalingFactor
+                              compressionQuality:compressionQuality];
```

Comment:
do we need to recompress the original image after it has been already compressed by XCTest? Wouldn't this be a waste of resources?

Reply:
Yes, otherwise the size keeps roughly the same.

Reply:
I removed the additional setting for compression quality and now we use the same value for screenshot taking and when we create the jpeg after scaling